### PR TITLE
(fix) 03-1995: Fix patient conditions sorting on "onSetDate".

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
- This PR fixes the sorting bug we had when we sort on `onSetDate` on the patient chart conditions table. This was reported by the `faimer team`. The date string we show on the UI was not properly sorting since it is just alphabetical, so there was need to specify the sortKey which is the actual dateTime string.

## Screenshots
Screenshot 2023-03-27 at 00 58 50

- Loom recording
https://www.loom.com/share/10a1246a12d74fc8ba7a02e7a3d45270

## Related Issue
- https://issues.openmrs.org/browse/O3-1995

### CODE REVIEW

File: packages/esm-patient-conditions-app/src/conditions/utils.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Enables a comparison of arbitrary values with support for undefined/null.
+ * Requires the `<` and `>` operators to return something reasonable for the provided values.
+ */
+export function compare<T>(x?: T, y?: T) {
```
Comment (vasharma05) [inline packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: You didn't consider the case for `--`

Comment (hadijahkyampeire) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: Let me check that, but I was handling it as an empty string.

Comment (hadijahkyampeire) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: @vasharma05 actually this compare function does not accept the `emdash` values, but I am handling such cases as undefined. Do you have some pointers on how you would want the `emdash` to be handled?

Comment (hadijahkyampeire) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: The `parseDate` will return a date string and that fails the `typescript` flags it as not being a number.

Comment (ibacher) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: `parseDate()` returns a JS Date object...

Comment (ibacher) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: So:

```ts
  if (x === y || !x && !y) {
    return 0;
  } else if (!x) {
    return -1;
  } else if (!y) {
    return 1;
  } else {
    const xDate = parseDate(x);
    const yDate = parseDate(y);
    
    if (xDate === yDate) {
       return 0;
    }

    return xDate < yDate ? -1 : 1;
  }
```

Comment (hadijahkyampeire) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts:5]: Thanks @ibacher

**Review by ayush-AI — COMMENTED** [review]
Hello @hadijahkyampeire, thanks you for this PR. In the PR title, mentioned issue should be O3 -1995 instead of 03 -1995. Please fix that.

File: packages/esm-patient-conditions-app/src/conditions/utils.ts (hunk 1 above)
Comment (vasharma05) [reply packages/esm-patient-conditions-app/src/conditions/utils.ts]: Ohh okay, I remembered `parseDate` returned millseconds, but forgot about the JS Date object.
Thanks Ian for the correction!

**Review by vasharma05 — APPROVED** [review]
Thanks for the fix @hadijahkyampeire !

**Pull request merged**